schematron/v2.1.0/tests/test_L000_Prelim_Analysis.py

```diff
@@ -0,0 +1,85 @@
+import os
+
+from lxml import etree
+
+from tools.validate_sch import validate_schematron
+from conftest import failures_by_role, AssertFailureRolesMixin, golden_tree, remove_element, SCH_DIR
+
+class TestL000PrelimAnalysis(AssertFailureRolesMixin):
+    schematron = os.path.join(SCH_DIR, 'L000_Prelim_Analysis.sch')
+    golden_file = os.path.join(SCH_DIR, 'golden_files', 'L000_Prelim_Analysis.xml')
+
+    def test_golden_file_is_valid(self):
+        # -- Act
+        failures = validate_schematron(self.schematron, self.golden_file)
+
+        # -- Assert
+        self.assert_failure_roles(failures, {})
+
+    def test_fails_when_measured_scenario_missing(self):
+        # -- Setup
+        tree = golden_tree('L000_Prelim_Analysis')
+
+        # verify it's valid
+        failures = validate_schematron(self.schematron, tree)
+        self.assert_failure_roles(failures, {})
+
+        # remove the measured scenario
+        tree = remove_element(tree, '//auc:Scenario[auc:ScenarioType/auc:CurrentBuilding/auc:CalculationMethod/auc:Measured]')
+        
+        # -- Act 
+        failures = validate_schematron(self.schematron, tree)
+
+        # -- Assert
+        self.assert_failure_roles(failures, {'ERROR': 3})
+
+    def test_fails_when_measured_scenario_not_linked_to_building(self):
+        # -- Setup
+        tree = golden_tree('L000_Prelim_Analysis')
+
+        # verify it's valid
+        failures = validate_schematron(self.schematron, tree)
+        self.assert_failure_roles(failures, {})
+
+        # remove the measured scenario
+        tree = remove_element(tree, '//auc:Scenario[auc:ScenarioType/auc:Benchmark]/auc:LinkedPremises/auc:Building/auc:LinkedBuildingID')
+
+        # -- Act 
+        failures = validate_schematron(self.schematron, tree)
+
+        # -- Assert
+        self.assert_failure_roles(failures, {'ERROR': 1})
+
+    def test_fails_when_benchmark_scenario_missing(self):
+        # -- Setup
+        tree = golden_tree('L000_Prelim_Analysis')
+
+        # verify it's valid
+        failures = validate_schematron(self.schematron, tree)
+        self.assert_failure_roles(failures, {})
+
+        # remove the benchmark scenario
+        tree = remove_element(tree, '//auc:Scenario/auc:ScenarioType/auc:Benchmark')
+
+        # -- Act 
+        failures = validate_schematron(self.schematron, tree)
+
+        # -- Assert
+        self.assert_failure_roles(failures, {'ERROR': 2})
```

Comment:
This I think is perfect example of why asserting the expected error messages would be helpful.  We remove the `auc:Benchmark` element.  Now, to me, we know longer have `count(auc:Scenario[auc:ScenarioType/auc:Benchmark]) &gt; 1`, which I think would be the only error?

Reply:
updated 👍